picard/util/checkupdate.py

```diff
@@ -0,0 +1,181 @@
+
+        Args:
+            show_always: Boolean value indicating whether the results dialog
+                should be shown even when there is no update available.
+            update_level: Determines what type of updates to check.  Options are:
+                0 = only stable release versions are checked.
+                1 = stable and beta releases are checked.
+                2 = stable, beta and dev releases are checked.
+
+        Returns:
+            none.
+
+        Raises:
+            none.
+        '''
+        self._show_always = show_always
+        self._update_level = update_level
+
+        if self._available_versions['stable']['tag']:
+            # Release information already acquired from specified website api.
+            self._display_results()
+        else:
+            # Gets list of releases from specified website api.
+            self._query_available_updates(callback=callback)
+
+    @property
+    def available_versions(self):
+        '''Provide a list of the latest version tuples for each update type.'''
+        return self._available_versions
+
+    def _query_available_updates(self, callback=None):
+        '''Gets list of releases from specified website api.'''
+        log.debug(_("Getting release information from {host_url}.".format(host_url=VERSIONS_API['host'],)))
+        self.tagger.webservice.get(
+            VERSIONS_API['host'],
+            VERSIONS_API['port'],
+            VERSIONS_API['endpoint'],
+            partial(self._releases_json_loaded, callback=callback),
+            parse_response_type=None,
+            priority=True,
+            important=True
+        )
+
+    def _releases_json_loaded(self, response, reply, error, callback=None):
+        '''Processes response from specified website api query.'''
+        if error:
+            log.error(_("Error loading releases list: {error_message}").format(error_message=reply.errorString(),))
+            if self._show_always:
+                QMessageBox.information(
+                    self._parent,
+                    _("Picard Update"),
+                    _("Unable to retrieve the latest version information."),
+                    QMessageBox.Ok, QMessageBox.Ok)
+        else:
+            self._available_versions = load_json(response)['versions']
+            for key in self._available_versions.keys():
+                log.debug("Version key '{version_key}' --> {version_information}".format(
+                    version_key=key, version_information=self._available_versions[key],))
+            self._display_results()
+        if callback:
+            callback(not error)
+
+    def _display_results(self):
+        # Display results to user.
+        key = ''
+        high_version = PICARD_VERSION
+        for test_key in PROGRAM_UPDATE_LEVELS:
+            if self._update_level >= PROGRAM_UPDATE_LEVELS[test_key]['level'] and  compare_version_tuples(high_version, self._available_versions[test_key]['version']) > 0:
```

Comment:
Two spaces instead of one just before `compare_version_tuples` and i'd wrap it (with parenthesis) since it's quite long plus a bit of variable caching:

```python
test_version = self._available_versions[test_key]['version']
if (self._update_level >= PROGRAM_UPDATE_LEVELS[test_key]['level']
    and compare_version_tuples(high_version, test_version):
        key = test_key
        high_version = test_version
```

Reply:
Good point on the caching and such.  I think I'll still need to check if the return value from `compare_version_tuples()` > 0 because the function could return -1 which (I think) would be evaluated as True, and I only want to continue processing if the value is 1.